language/vm/src/file_format.rs

```diff
@@ -1615,6 +1646,9 @@ pub struct CompiledModuleMut {
     pub struct_defs: Vec<StructDefinition>,
     /// Function defined in this module.
     pub function_defs: Vec<FunctionDefinition>,
+
+    /// Friend declarations of this module.
+    pub friend_decls: Vec<FriendDeclaration>,
```

Comment:
I'm not super convinced that we should be using the term "friend" any more. I think at the bytecode level we should just be more explicit and descriptive. Pulling into a single term like "friend" seems useful only if we need to refer to this in a bunch of different places. Which I don't think is true?

So perhaps just 
```
pub protected_access: Vec<ModuleHandleIndex>
```

Reply:
I am fine with either naming, but I guess we may actually need refer to the "friend" concept across the stack? like file format, bytecode verifier, IR compiler, source compiler, prover, VM loader, etc?

Reply:
I strongly agree--I think `friend` is a bad term for us to use because it implies symmetry, but the language feature is not symmetric. Perhaps we should elevate this terminology question to a broader forum, but I think something like "client" or "patron" that connotes anti-symmetry would be good.

Reply:
Agree with Sam, and I would go one step further: I think we don't need to pin ourselves to a term like "friend" or "client" at all because the usecase is narrow. 

My vote for the syntax is 
```
protected  { 0x1::A, 0x2::B, 0x3::C }
```
No need for another keyword and it makes it hopefully more clear what is going on. If I saw "friend" here I would have no initial guess that it relates to `protected`

Reply:
After some internal discussion, we decided to go with the `friend` keyword and using `public(friend)` as the visibility modifier.

Reply:
Do we want `friend` for `friends`? :P 

Seriously, I don't know. 

I think `public(friend)` reads better. But `friends { 0x1::A, 0x2::B }` reads better

Reply:
I think `public(friend)` reads better and aligns with `public(script)`.

An alternative of friend declaration would be
```
friend 0x1::A;
friend 0x2::B;
```
similar to how we deal with `use`.

Reply:
I'll vote for the plural. `public(friends)` reads better to me. If you compare it to Rust's `pub(crate)`, the code is only in one crate so it makes sense to use singular `crate` -- the visibility is public within the one crate. In our case, `public(friends)` means that the function is public to friend modules, and since there can be multiple friends, `public(friends)` is the right way to say that.

Reply:
...but after reading Meng's response, I would also feel OK about the singular `friend`. The comparison to `public(script)` is better than my comparison to Rust's `pub(crate)`

Reply:
AH! very clever! I like it :) 

We should maybe add coalescing to this and to use. So you can do `friend 0x2::{A, B};`

We should do this for use too. But its okay if it doesn't make it in the first PR